### PR TITLE
Path entity

### PR DESCRIPTION
**Pull request by garydgregory**

### CODE REVIEW

File: httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/PathEntity.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * A self contained, repeatable entity that obtains its content from a path.
+ */
+@Contract(threading = ThreadingBehavior.IMMUTABLE_CONDITIONAL)
+public class PathEntity extends AbstractHttpEntity {
+
+    private final Path path;
+
+    public PathEntity(final Path path, final ContentType contentType, final String contentEncoding) {
+        super(contentType, contentEncoding);
+        this.path = Args.notNull(path, "File");
```
Comment (michael-o) [inline httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/PathEntity.java]: It should be "path".

File: httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+/**
+ * Unit tests for {@link PathEntity}.
+ */
+public class TestPathEntity {
+
+    @Test
+    public void testBasics() throws Exception {
+        final Path tmpPath = Files.createTempFile("testfile", ".txt");
+        tmpPath.toFile().deleteOnExit();
+        try (final PathEntity httpentity = new PathEntity(tmpPath, ContentType.TEXT_PLAIN)) {
```
Comment (michael-o) [inline httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java]: Typo `httpEntity`

File: httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            Assert.assertEquals(Files.size(tmpPath), httpentity.getContentLength());
```
Comment (michael-o) [inline httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java]: How does this test make sense if the file is empty?

File: httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            try (final InputStream content = httpentity.getContent()) {
+                Assert.assertNotNull(content);
+            }
+            Assert.assertTrue(httpentity.isRepeatable());
+            Assert.assertFalse(httpentity.isStreaming());
+            if (!Files.deleteIfExists(tmpPath)) {
```
Comment (michael-o) [inline httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java]: Why is this in the try clause?
Why do you delete this file when you say: `tmpPath.toFile().deleteOnExit();`?

File: httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        tmpPath.toFile().deleteOnExit();
+
+        try (final OutputStream outStream = Files.newOutputStream(tmpPath)) {
+            outStream.write(0);
+            outStream.write(1);
+            outStream.write(2);
+            outStream.write(3);
+        }
+
+        try (final PathEntity httpentity = new PathEntity(tmpPath, ContentType.TEXT_PLAIN)) {
```
Comment (michael-o) [inline httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java]: case

File: httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            final ByteArrayOutputStream out = new ByteArrayOutputStream();
+            httpentity.writeTo(out);
+            final byte[] bytes = out.toByteArray();
+            Assert.assertNotNull(bytes);
+            Assert.assertEquals(Files.size(tmpPath), bytes.length);
+            for (int i = 0; i < 4; i++) {
+                Assert.assertEquals(i, bytes[i]);
+            }
+            if (!Files.deleteIfExists(tmpPath)) {
```
Comment (michael-o) [inline httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java]: Same here.

File: httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/PathEntity.java (hunk 1 above)
Comment (garydgregory) [reply httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/PathEntity.java]: Fixed.

File: httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java (hunk 2 above)
Comment (garydgregory) [reply httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java]: Fixed. Note that a lot of tests make this mistake. That's what I get for copy & pasting ;-)

File: httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java (hunk 3 above)
Comment (garydgregory) [reply httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java]: Added comment.

File: httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java (hunk 4 above)
Comment (garydgregory) [reply httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java]: Added comment.

File: httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java (hunk 5 above)
Comment (garydgregory) [reply httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java]: Fixed. Note that a lot of tests make this mistake. That's what I get for copy & pasting ;-)

File: httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java (hunk 6 above)
Comment (garydgregory) [reply httpcore5/src/test/java/org/apache/hc/core5/http/io/entity/TestPathEntity.java]: Added comment.

**Comment by garydgregory** [comment]
Closing in favor of https://github.com/apache/httpcomponents-core/pull/187

**Pull request closed without merging**